### PR TITLE
add sticky decision execution

### PR DESCRIPTION
**Pull request by yiminc**
this is part 1 of the sticky execution change.
this change added sticky task list and DecisionScheduleToStartTimeout for sticky task.
This part does not include partial history for sticky task, which will be worked on next.

### CODE REVIEW

File: service/history/historyEngine.go [hunk 1]
```diff
@@ -1064,13 +1083,19 @@ Update_History_Loop:
...
 		var transferTasks []persistence.Task
+		var timerTasks []persistence.Task
 		if !msBuilder.HasPendingDecisionTask() {
 			newDecisionEvent, _ := msBuilder.AddDecisionTaskScheduledEvent()
 			transferTasks = []persistence.Task{&persistence.DecisionTask{
 				DomainID:   domainID,
 				TaskList:   *newDecisionEvent.DecisionTaskScheduledEventAttributes.TaskList.Name,
 				ScheduleID: *newDecisionEvent.EventId,
 			}}
+			if len(msBuilder.executionInfo.StickyTaskList) > 0 {
```
Comment (samarabbas) [inline service/history/historyEngine.go]: Looks like this is repeated multiple times, can you add a helper method?

File: service/history/historyEngine.go [hunk 2]
```diff
@@ -580,6 +580,14 @@ Update_History_Loop:
 		var continueAsNewBuilder *mutableStateBuilder
 		hasDecisionScheduleActivityTask := false
 
+		if request.StickyAttributes == nil {
```
Comment (samarabbas) [inline service/history/historyEngine.go]: Can we add a metric on requests with sticky tasklists

File: service/history/timerQueueProcessor.go [hunk 3]
```diff
@@ -742,15 +745,31 @@ Update_History_Loop:
...
+				// Add a decision task timeout event.
+				timeoutEvent := msBuilder.AddDecisionTaskTimedOutEvent(scheduleID, di.StartedID)
+				if timeoutEvent == nil {
+					// Unable to add DecisionTaskTimedout event to history
+					return &workflow.InternalServiceError{Message: "Unable to add DecisionTaskTimedout event to history."}
+				}
 
+				scheduleNewDecision = true
+			}
+		case int(workflow.TimeoutTypeScheduleToStart):
```
Comment (samarabbas) [inline service/history/timerQueueProcessor.go]: Can we have a metric on each of these timeouts?

Comment (yiminc) [reply service/history/timerQueueProcessor.go]: added t.metricsClient.IncCounter(metrics.TimerTaskDecisionTimeoutScope, metrics.ScheduleToStartTimeoutCounter) and t.metricsClient.IncCounter(metrics.TimerTaskDecisionTimeoutScope, metrics.StartToCloseTimeoutCounter)

File: service/history/historyEngine.go (hunk 1 above)
Comment (yiminc) [reply service/history/historyEngine.go]: added mutableStateBuilder.isStickyTaskListEnabled()

File: service/history/historyEngine.go (hunk 2 above)
Comment (yiminc) [reply service/history/historyEngine.go]: added e.metricsClient.IncCounter(metrics.HistoryRespondDecisionTaskCompletedScope, metrics.CompleteDecisionWithStickyEnabledCounter)
and e.metricsClient.IncCounter(metrics.HistoryRespondDecisionTaskCompletedScope, metrics.CompleteDecisionWithStickyDisabledCounter)

File: common/persistence/cassandraPersistence.go [hunk 4]
```diff
@@ -1982,6 +1989,10 @@ func createWorkflowExecutionInfo(result map[string]interface{}) *WorkflowExecuti
 			info.CancelRequested = v.(bool)
 		case "cancel_request_id":
 			info.CancelRequestID = v.(string)
+		case "sticky_task_list":
```
[the remark this answers is not in the document]
Comment (yiminc) [reply common/persistence/cassandraPersistence.go]: which unit test?

**Pull request merged**